zio-http/src/main/scala/zio/http/Client.scala

```diff
@@ -29,21 +40,152 @@ trait Client {
+    copy(sslOption = Some(ssl))
+
+  protected def requestInternal(
+    body: Body,
+    headers: Headers,
+    hostOption: Option[String],
+    method: Method,
+    pathPrefix: Path,
+    portOption: Option[Int],
+    queries: Map[String, Chunk[String]],
+    sslOption: Option[ClientSSLOptions],
+    version: Version,
+  ): ZIO[Any, Throwable, Response]
+
+  private def copy(
+    headers: Headers = headers,
+    hostOption: Option[String] = hostOption,
+    pathPrefix: Path = pathPrefix,
+    portOption: Option[Int] = portOption,
+    queries: Map[String, Chunk[String]] = queries,
+    sslOption: Option[ClientSSLOptions] = sslOption,
+  ): Client =
+    Client.Proxy(self, headers, hostOption, pathPrefix, portOption, queries, sslOption)
 }
 
 object Client {
+
+  private final case class Proxy(
+    client: Client,
+    headers: Headers,
+    hostOption: Option[String],
+    pathPrefix: Path,
+    portOption: Option[Int],
+    queries: Map[String, Chunk[String]],
+    sslOption: Option[ClientSSLOptions],
+  ) extends Client {
+
+    def requestInternal(
+      body: Body,
+      headers: Headers,
+      hostOption: Option[String],
+      method: Method,
+      path: Path,
+      portOption: Option[Int],
+      queries: Map[String, Chunk[String]],
+      sslOption: Option[ClientSSLOptions],
+      version: Version,
+    ): ZIO[Any, Throwable, Response] =
+      client.requestInternal(body, headers, hostOption, method, path, portOption, queries, sslOption, version)
+
+    def socket[R](
+      url: String,
+      app: SocketApp[R],
+      headers: Headers = Headers.empty,
+      addZioUserAgentHeader: Boolean = false,
+    )(implicit unsafe: Unsafe): ZIO[R with Scope, Throwable, Response] =
+      client.socket(url, app, headers, addZioUserAgentHeader)
+
+  }
+
   final case class ClientLive(
     settings: ClientConfig,
     rtm: HttpRuntime[Any],
     cf: JChannelFactory[JChannel],
     el: JEventLoopGroup,
   ) extends Client
       with ClientRequestEncoder {
+    val headers: Headers                    = Headers.empty
+    val hostOption: Option[String]          = None
+    val pathPrefix: Path                    = Path.empty
+    val portOption: Option[Int]             = None
+    val queries: Map[String, Chunk[String]] = Map.empty
+    val sslOption: Option[ClientSSLOptions] = None
+
+    def requestInternal(
+      body: Body,
+      headers: Headers,
+      hostOption: Option[String],
+      method: Method,
+      path: Path,
+      portOption: Option[Int],
+      queries: Map[String, Chunk[String]],
+      sslOption: Option[ClientSSLOptions],
+      version: Version,
+    ): ZIO[Any, Throwable, Response] = {
 
-    override def request(request: Request): ZIO[Any, Throwable, Response] =
-      requestAsync(request, settings)
+      for {
+        host     <- ZIO.fromOption(hostOption).orElseFail(new IllegalArgumentException("Host is required"))
+        port     <- ZIO.fromOption(portOption).orElseSucceed(80)
```

Comment:
Maybe we should default to `443` if `sslOption` is defined

Reply:
I agree, @adamgfraser can you do a followup?